### PR TITLE
Validating openAI key

### PR DESCRIPTION
**Pull request by WilliamEspegren**
An easy way to check if the key is a valid API key is to check if it is 51 characters long. If it isn't, it isn't a OpenAI key. We still don't really know if the key is valid, but it helps with people typing/copying the wrong (amount) of characters etc...

Edit: Added another commit that makes a "test" request to the openAI endpoint to confirm that it is a valid token.

### CODE REVIEW

File: interpreter/interpreter.py [hunk 1]
```diff
@@ -393,51 +393,57 @@ def verify_api_key(self):
...
+                answers = inquirer.prompt(questions)
+                chosen_param = answers['param']
 
+                # THIS is more in line with the future. You just say the model you want by name:
+                self.model = models[chosen_param]
+                self.local = True
+                
+                return
 
+            elif len(response) != 51:
```
[the remark this answers is not in the document]
Comment (WilliamEspegren) [reply interpreter/interpreter.py]: I have added another commit that makes a "test" request to the openAI endpoint to confirm that the key is actually valid 👍

**Comment by WilliamEspegren** [comment]
> @WilliamEspegren this is an awesome PR ! We can expose this as a util on litellm, `litellm.validate_openai_key()` 
> 
> benefits of this: it will keep the OI code base simpler. Since LiteLLM focuses on Input/output of llms this function feels like a natural fit there 
> 
> open to your thoughts on this 

Great idea! Will look into how to create a util in LiteLLM, since I'm new to it 🤝

**Comment by WilliamEspegren** [comment]
> 

Great idea! Will look into how to create a util in LiteLLM, since I'm new to it 🤝

**Comment by ishaan-jaff** [comment]
happy to work on this with you - here's a link to our discord: https://discord.com/invite/wuPM9dRgDw 
we can co-ordinate there

**Comment by WilliamEspegren** [comment]
With @ishaan-jaff and @krrishdholakia help we discussed and refined a util in LiteLLM that checks if the token provided by the user is valid or not. In the future open-interpreter can support other models, because LiteLLM checks the validity of many LLM's tokens.

**Comment by krrishdholakia** [comment]
🎉 teamwork

**Comment by jordanbtucker** [comment]
So, I'm on the fence about this. There's no guarantee that OpenAI will always produce 51 character API keys. They could change that at any moment. So this is possibly something that breaks moving forward out of the blue and would require maintenance.

Is key validation something that has really been a problem for users?

**Comment by WilliamEspegren** [comment]
Maybe not an issue that users are facing, we can remove the 51 character check and only check using liteLLM?

**Comment by jordanbtucker** [comment]
I think it would be prudent to remove the 51 character check.

As for using LiteLLM, I checked the code and all it does it send a message to check if it gets back an `AuthenticationError`. The problem with that is, if there is no AuthenticationError, then we're just wasting tokens, which cost money.

Sorry for being a hardass. I think the best option is to have interpreter check to see if it's getting AuthenticationErrors back from LiteLLM and display a friendly error message if so. But there's no need to do that pre-emptively.

**Comment by WilliamEspegren** [comment]
Okey, perfect. Just an idea 👍

**Comment by haseeb-heaven** [comment]
This is great way of checking, But if you already make request with Invalid Key the OpenAI already sends you message back saying `"API Key is Invalid"` so is this PR really necessary ?

**Comment by WilliamEspegren** [comment]
@haseeb-heaven Thank you! The point of this PR was to check whether the key provided (no matter what LLM) is valid or not, before running open-interpreter. Now, when you enter an invalid key, open-interpreter just returns an error and crashes. I guess we can try the completion and if that fails just ask for the key again?

**Comment by haseeb-heaven** [comment]
> @haseeb-heaven Thank you! The point of this PR was to check whether the key provided (no matter what LLM) is valid or not, before running open-interpreter. Now, when you enter an invalid key, open-interpreter just returns an error and crashes. I guess we can try the completion and if that fails just ask for the key again?

Great then but there are two ways to check that.
1.Offline - Check keys length and syntax before adding them.
2.Online - Send a dummy request to OpenAI with that key and check the response ensuring Key is valid.

So i think we should add both ways of checking this is effective.

**Comment by WilliamEspegren** [comment]
> Great then but there are two ways to check that. 1.Offline - Check keys length and syntax before adding them. 2.Online - Send a dummy request to OpenAI with that key and check the response ensuring Key is valid.
> 
> So i think we should add both ways of checking this is effective.

Checking the lenght offline was my first commit, but as @jordanbtucker wrote: 
>So, I'm on the fence about this. There's no guarantee that OpenAI will always produce 51 character API keys.

and

>I think it would be prudent to remove the 51 character check.

So then I changed it to use LiteLLM that sends a dummy request, but as @jordanbtucker wrote, that wastes tokens.

The thing we can do now, is accept any input when the user is asked for their API key, and then, when the user writes a prompt to open-interpreter (with an invalid API key) we can catch that error and ask the user for another key.

**Comment by haseeb-heaven** [comment]
> > Great then but there are two ways to check that. 1.Offline - Check keys length and syntax before adding them. 2.Online - Send a dummy request to OpenAI with that key and check the response ensuring Key is valid.
> > So i think we should add both ways of checking this is effective.
> 
> Checking the lenght offline was my first commit, but as @jordanbtucker wrote:
> 
> > So, I'm on the fence about this. There's no guarantee that OpenAI will always produce 51 character API keys.
> 
> and
> 
> > I think it would be prudent to remove the 51 character check.
> 
> So then I changed it to use LiteLLM that sends a dummy request, but as @jordanbtucker wrote, that wastes tokens.
> 
> The thing we can do now, is accept any input when the user is asked for their API key, and then, when the user writes a prompt to open-interpreter (with an invalid API key) we can catch that error and tell the user that the key was invalid, then ask the user for the key again.

Yes that is the 2nd solution to get OpenAI to verify this Key, but if you're sure that how api key is constructed and how to validate it then 1st solution as well but we need to have set of valid and invalid keys for OpenAI and then compare them with the user input key for validity.

And yes if we send request to server for validation of key it would be waste of token but it will be valid way of checking the api key.

**Comment by ericrallen** [comment]
Hey there, folks!

I'm going to close this one as stale for now, but if we think it's still important and valid, please feel encouraged to resolve the conflicts (which will probably require updating to put this in the right spot due to the codebase's recent refactor) and reopen it.

**Pull request closed without merging**